### PR TITLE
adopt chat agent in terminal chat, fix a few bugs

### LINKED ISSUE #214208: Fully adopt agents for terminal inline chat
The use of inline chat for terminals relies on a few deprecated methods and code paths. Instead, terminal chat should solely rely on agents, like notebook and editor chat. This means

* register an agent for the terminal location (from the extension)
* inside the extension dispatch the agent invocation
* in the renderer create the widget for the terminal locations
* do away with custom "make request" actions and rely on the chat widget, e.g make sure the default send invokes `ChatService#sendRequest`
* remove more custom logic, like history management, placeholder text etc pp

### PR DESCRIPTION
**Pull request by meganrogge**
fixes #214208
fixes https://github.com/microsoft/vscode-copilot/issues/5246
fixes https://github.com/microsoft/vscode-copilot/issues/5228

### CODE REVIEW

File: src/vs/workbench/contrib/terminalContrib/chat/browser/terminalChatActions.ts [hunk 1]
```diff
@@ -283,43 +279,13 @@ registerActiveXtermAction({
 	}
 });
 
-registerActiveXtermAction({
-	id: TerminalChatCommandId.ViewInChat,
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminalContrib/chat/browser/terminalChatActions.ts:287]: Is this now redundant? We want to keep view in chat

**Comment by Tyriar** [comment]
Follow ups are being shown on top which obscures the content:

image

**Comment by Tyriar** [comment]
Pressing escape now seems to clear the chat but not hide it:

image

**Comment by meganrogge** [comment]
> Pressing escape now seems to clear the chat but not hide it:
> 
> image

Cannot repro on mac 🤔

File: src/vs/workbench/contrib/terminalContrib/chat/browser/terminalChatActions.ts (hunk 1 above)
Comment (meganrogge) [reply src/vs/workbench/contrib/terminalContrib/chat/browser/terminalChatActions.ts]: added it back 👍🏼

File: src/vs/workbench/contrib/inlineChat/browser/inlineChatWidget.ts [hunk 2]
```diff
@@ -448,12 +448,12 @@ export class InlineChatWidget {
 		if (!viewModel) {
 			return undefined;
 		}
-		for (const item of viewModel.getItems()) {
-			if (isResponseVM(item)) {
-				return viewModel.codeBlockModelCollection.get(viewModel.sessionId, item, codeBlockIndex)?.model;
-			}
+		const items = viewModel.getItems().filter(i => isResponseVM(i));
```
Comment (meganrogge) [inline src/vs/workbench/contrib/inlineChat/browser/inlineChatWidget.ts:451]: we use this to identify which actions should be shown - run/insert/discard or view in chat. formerly, there was only one viewModel item. now, bc of the underlying `ChatWidget`, there are many, so we want to look at the last one.

File: src/vs/workbench/contrib/terminalContrib/chat/browser/terminalChatWidget.ts [hunk 3]
```diff
@@ -155,11 +155,15 @@ export class TerminalChatWidget extends Disposable {
 		if (!terminalWrapperHeight) {
 			return;
 		}
-		if (top > terminalWrapperHeight - widgetHeight) {
+		if (top > terminalWrapperHeight - widgetHeight && terminalWrapperHeight - widgetHeight > 0) {
```
Comment (meganrogge) [inline src/vs/workbench/contrib/terminalContrib/chat/browser/terminalChatWidget.ts:158]: prevents the top of the widget from going above the terminal container

File: src/vs/workbench/contrib/terminalContrib/chat/browser/terminalChatWidget.ts [hunk 4]
```diff
@@ -155,11 +155,15 @@ export class TerminalChatWidget extends Disposable {
...
 			this._setTerminalOffset(top - (terminalWrapperHeight - widgetHeight));
 		} else {
 			this._setTerminalOffset(undefined);
 		}
+		if (terminalWrapperHeight - widgetHeight < 0) {
```
Comment (meganrogge) [inline src/vs/workbench/contrib/terminalContrib/chat/browser/terminalChatWidget.ts:163]: prevents the height from outgrowing the terminal container

File: src/vs/workbench/contrib/terminalContrib/chat/browser/terminalChatWidget.ts [hunk 5]
```diff
@@ -155,11 +155,15 @@ export class TerminalChatWidget extends Disposable {
...
+			this._dimension = new Dimension(this._dimension!.width, terminalWrapperHeight - top - 20);
```
Comment (meganrogge) [inline src/vs/workbench/contrib/terminalContrib/chat/browser/terminalChatWidget.ts:164]: 20 looks nice / not cramped IMO 

Screenshot 2024-06-07 at 12 46 33 PM

Comment (meganrogge) [reply src/vs/workbench/contrib/terminalContrib/chat/browser/terminalChatWidget.ts:164]: note this is in oss not using the dev ext host window - there will be no follow ups

File: src/vs/workbench/contrib/terminalContrib/chat/browser/terminalChatActions.ts [hunk 6]
```diff
@@ -366,25 +359,43 @@ registerActiveXtermAction({
...
+});
+
+registerActiveXtermAction({
+	id: TerminalChatCommandId.NextFromHistory,
+	title: localize2('nextFromHitory', 'Next From History'),
+	precondition: ContextKeyExpr.and(
+		TerminalChatContextKeys.focused,
+	),
+	keybinding: {
+		when: ContextKeyExpr.and(TerminalChatContextKeys.focused),
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminalContrib/chat/browser/terminalChatActions.ts:389]: Don't need and

File: src/vs/workbench/contrib/terminalContrib/chat/browser/terminalChatActions.ts [hunk 7]
```diff
@@ -366,25 +359,43 @@ registerActiveXtermAction({
...
 	),
-	icon: Codicon.report,
-	menu: [{
-		id: MENU_TERMINAL_CHAT_WIDGET_FEEDBACK,
-		when: ContextKeyExpr.and(TerminalChatContextKeys.responseContainsCodeBlock.notEqualsTo(undefined), TerminalChatContextKeys.responseSupportsIssueReporting),
-		group: 'inline',
-		order: 3
-	}],
+	keybinding: {
+		when: ContextKeyExpr.and(TerminalChatContextKeys.focused),
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminalContrib/chat/browser/terminalChatActions.ts:368]: Don't need and

File: src/vs/workbench/contrib/terminalContrib/chat/browser/terminalChatController.ts [hunk 8]
```diff
@@ -81,39 +76,41 @@ export class TerminalChatController extends Disposable implements ITerminalContr
...
 
 	private readonly _model: MutableDisposable<ChatModel> = this._register(new MutableDisposable());
 
 	get scopedContextKeyService(): IContextKeyService {
 		return this._chatWidget?.value.inlineChatWidget.scopedContextKeyService ?? this._contextKeyService;
 	}
 
+	private _sessionCtor: CancelablePromise<void> | undefined;
+	private static _storageKey = 'terminal-inline-chat-history';
+	private static _promptHistory: string[] = [];
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminalContrib/chat/browser/terminalChatController.ts:88]: Can these move up near the other statics?

**Review by Tyriar — APPROVED** [review]

**Pull request merged**